### PR TITLE
amd qwen35 optimize fused_l2norm_qk

### PR DESCRIPTION
**Pull request by hxy0118**
- kernel: replace x/sqrt(var) with x*rsqrt(var) — div is multi-instruction rcp+NR on AMD/NV, broadcasting it across BT*BD becomes the bottleneck
- dispatcher: drop the T<=128 gate, make T/D runtime args so the BT-tiled path no longer recompiles per-shape; num_warps 8->2

Measured on MI308X, 9B prefill T=122936 D=128 bf16:
  before  842us  ->  after 49us  (17x), beats sgl 100us by 2.0x

### CODE REVIEW

**Comment by LLLLKKKK** [comment]
## 🤖 RTP-LLM PR #920 Code Review

**amd qwen35 optimize fused_l2norm_qk** by @hxy0118 (温华)
**Commit:** `93b2f806` | **Files:** 3 | **+188 / -4**

### 目标
将 Qwen3.5 中 Q/K 的两次独立 L2 norm 调用合并为单次 fused kernel launch，减少 kernel launch 开销和重复内存访问。

### 关键变更
1. **`fused_l2norm_qk_kernel`** (BT-tiled) — D≤512 时使用，BT=16 行一起处理，2 warps。同一 kernel 内先处理 Q 再处理 K，共享 row/col mask 计算
2. **`fused_l2norm_qk_kernel1`** (per-row) — D>512 时使用，每行一个 program，8 warps
3. **rsqrt 优化** — 使用 `rsqrt + mul` 替代 `div`，利用硬件单指令 rsqrt，注释说明 ~2.3x 加速
4. **测试** — 新增 `TestFusedL2NormQK`，验证与两次独立 `l2norm_fwd` 的数值一致性，覆盖多种 T/D/dtype 组合

### ✅ 优点
- 优化思路清晰：fuse 两次 kernel launch 为一次，减少 launch overhead
- rsqrt 替代 div 的性能注释详细且有实测数据
- 测试覆盖全面：parity check、dispatch 分支、shape/dtype assert、多种 prefill 长度（16~65536）
- BT-tiled vs per-row 的 dispatch 策略有性能分析注释

### 问题

🔵 **P2 (1)**:
1. **`fused_l2norm_qk_kernel1` 参数顺序** — `eps` 作为最后一个参数而非 `tl.constexpr` 之后的第一个，与 `fused_l2norm_qk_kernel` 的参数顺序不一致（后者 eps 在 T/D 之前）。不影响正确性但降低可读性。

### 结论

✅ **LGTM** — 干净的性能优化，数值正确性有充分测试保证。

**Review by LLLLKKKK — COMMENTED** [review]
## Code Review: amd qwen35 optimize fused_l2norm_qk

### Summary

Fuses two independent `l2norm_fwd(q)` and `l2norm_fwd(k)` calls into a single `fused_l2norm_qk(q, k)` kernel launch. Two new Triton kernels: BT-tiled (D<=512) and per-row (D>512). Key optimizations: `rsqrt + mul` instead of `div`, and T/D as runtime args to avoid recompilation across sequence lengths. Tests and ROCm BUILD target added.

### Findings

**[P2] `num_warps=2` for BT-tiled kernel may underperform on NVIDIA** (`l2norm.py`)
The BT-tiled kernel uses `num_warps=2` while the original `l2norm_fwd_kernel` uses `num_warps=8`. Since `chunk.py` is platform-agnostic, this kernel replaces the NVIDIA path too. With BT=16 and BD up to 512, 2 warps may underutilize NVIDIA SMs. Was this benchmarked on NVIDIA (H20/A100) as well? Consider platform-aware tuning or `triton.autotune`.

**[P2] Misleading comment in test** (`test_l2norm.py`)
```python
# T<=128 走 BT-tiled kernel；其余走 (T,) per-row kernel
```
The dispatch condition is `D <= 512`, not `T <= 128`. Should be: `# D<=512 走 BT-tiled kernel；D>512 走 per-row kernel`.

**[P2] `k_out.view(q_shape_og)` uses q's shape** (`l2norm.py`)
```python
return q_out.view(q_shape_og), k_out.view(q_shape_og)
```
The assertion guarantees `q.shape == k.shape` so this is correct, but using `q_shape_og` for both is slightly confusing. Consider saving `k_shape_og` separately or adding a brief comment.

**[P2] Missing newline at end of BUILD file** (`test/BUILD`)
`\ No newline at end of file` — minor hygiene issue that causes noisy diffs.

**[P3] Test doesn't cover D<=512 / D>512 dispatch boundary** (`test_l2norm.py`)
Test cases are `D in [128, 256, 128, 128, 1024]`. Only `D=1024` exercises the per-row kernel. Consider adding `D=512` (boundary) and `D=513` to explicitly test the dispatch edge.

### Overall

Well-structured PR with sound optimization approach. The mathematical transformation (`rsqrt+mul` vs `div`) is correct. Main concern is whether `num_warps=2` was validated on NVIDIA since this is a shared code path. No blockers.

---
🤖 Automated review by [Claude Code](https://claude.ai/code)

**Comment by LLLLKKKK** [comment]
🤖 AI Code Review — PR #920

## PR 概述

**Title:** `amd qwen35 optimize fused_l2norm_qk`
**Author:** hxy0118
**规模:** 4(GitHub) files, +201/-4

### 核心目标

将 `chunk.py` 中对 q、k 的两次独立 `l2norm_fwd` 调用合并为一次 `fused_l2norm_qk` kernel launch，同时用 `rsqrt + mul` 替代 `div`（AMD/NV 上 div 是多指令 rcp+NR，广播到 BT*BD 成为瓶颈），并将 T/D 改为 runtime arg 消除 per-shape 重编译。MI308X 上 9B prefill 从 842us 降至 49us（17x）。

---

## 改动逻辑拆解

### GitHub 开源仓库变更（主要代码）

#### 1. l2norm.py — 新增 fused kernel 及 dispatcher

新增两个 Triton kernel：

- `fused_l2norm_qk_kernel`：BT-tiled 路径（D<=512），T/D 为 runtime arg，BT/BD 为 constexpr。每个 program 处理 BT 行的 q 和 k，用 `rsqrt(sum(x*x) + eps) * x` 替代原来的 `x / sqrt(var + eps)`。
- `fused_l2norm_qk_kernel1`：per-row 路径（D>512），每个 program 处理一行 q 和一行 k。

新增 dispatcher `fused_l2norm_qk(q, k, eps, output_dtype)`：
- 校验 q/k shape 和 dtype 一致
- D<=512 走 BT-tiled kernel（BT=16, num_warps=2）
- D>512 走 per-row kernel（num_warps=8）
- 返回 `(q_out, k_out)` tuple

关键设计决策：T 和 D 不再是 `tl.constexpr`，避免不同 batch size 触发重编译（原 `l2norm_fwd_kernel` 的 T 是 constexpr，每个新 T 值都要编译一次，70ms 编译 vs <1ms 执行）。

#### 2. chunk.py — 调用方适配

import 从 `l2norm_fwd` 改为 `fused_l2norm_qk`，`forward()` 中两行 `q = l2norm_fwd(q); k = l2norm_fwd(k)` 合并为 `q, k = fused_l2norm_qk(q, k)`。

#### 3. test/test_l2norm.py — 新增 TestFusedL2NormQK 测试类

- `_check_parity`：对比 fused 结果与两次独立 `l2norm_fwd` 的结果
- `test_parity_qwen3_next_shapes`：D=128, T 从 16 到 65536
- `test_parity_dispatch_branches`：覆盖 D<=512 和 D>512 两条路径，bf16/fp16/fp32 三种 dtype
- `test_shape_dtype_assert`：校验 shape/dtype 不匹配时的 assert

#### 4. test/BUILD — 新增 ROCm 测试 target

新增 `test_l2norm_rocm`，复用 `test_l2norm.py`，指定 `tags = ["rocm"]` 和 `exec_properties = {'gpu':'MI308X'}`。

---

## Checklist 检查结果

### 通用原则

#### 软件工程原则
| 检查项 | 结果 |
|--------|------|
| SRP | ✅ 单一职责：fused kernel + dispatcher + 调用方适配 |
| OCP | ✅ 新增 fused 函数，未修改原 `l2norm_fwd` 逻辑 |
| LSP | ✅ 不涉及继承 |
| ISP | ✅ 接口精简 |
| DIP | ✅ 不涉及 |
| DRY | ✅ fused kernel 内 q/k 处理逻辑相同但操作不同数据，无需抽取 |
| KISS | ✅ 实现直接，无过度抽象 |
| YAGNI | ✅ 只做了需要的优化 |

#### 架构审视
| 检查项 | 结果 |
|--------|------|
| 抽象边界 | ✅ fused kernel 放在 l2norm.py 中，与原有 kernel 同层 |
| 依赖方向 | ✅ chunk.py → l2norm.py，方向不变 |
| 状态完整性 | ✅ 无状态变更 |
| 错误语义 | ✅ shape/dtype 不匹配时 assert 失败，D 过大时 RuntimeError |
| 可观测性 | ✅ 不涉及 |
| 可演进性 | ✅ 原 `l2norm_fwd` 保留，不影响其他调用方 |
| 可运维性 | ✅ 不涉及 |

#### 测试
| 检查项 | 结果 |
|--------|------|
| 新功能有对应测试 | ✅ TestFusedL2NormQK 覆盖多种 shape/dtype/dispatch 路径 |
| 删除的测试有等价替代 | ✅ 未删除测试 |
| 边界 case 覆盖 | ✅ T=16（小）到 T=65536（大），D=128/256/1024，三种 dtype |
| 分布式改动有多卡测试 | ✅ 不涉及分布式 |

#### 代码质量与文档
| 检查项 | 结果 |
|--------|------|
| 无关改动分离 | ✅ 所有改动围绕 fused l2norm |
| mega-PR 拆分 | ✅ 单一功能 |
| Commit 原子性 | ✅ 单 commit，自包含 |
| Commit message 准确性 | ✅ 准确描述优化内容 |
| PR description 说明动机和设计 | ✅ 包含优化原理和性能数据 |
| 日志频率控制 | ✅ 不涉及 |

### 领域检查

#### A. 兼容性与配置 — 全部 ✅
#### B. 正确性与逻辑 — 全部 ✅
#### C. 线程安全与并发 — 全部 ✅
#### D. 性能 — 全部 ✅
#### E. 分布式 — 全部 ✅（不涉及）
#### F. 跨平台（ROCm/ARM） — 全部 ✅
#### G. 语言与框架特有 — 全部 ✅
#### H. 测试与 CI — 全部 ✅
#### I. 代码质量 — 全部 ✅

---

## Review 意见

### 问题

1. **`l2norm_fwd` 中旧 BT-tiled kernel 仍有 T 作为 constexpr 的重编译问题** [P2]

   `fused_l2norm_qk_kernel` 已将 T/D 改为 runtime arg 解决了重编译问题，但原 `l2norm_fwd` 中的 `l2norm_fwd_kernel` 仍然将 T 声明为 `tl.constexpr`（第 60 行），且 `l2norm_fwd` 仍被 `L2NormFunction.forward` 调用。如果有其他路径（非 fused）调用 `l2norm_fwd`，仍会遇到 per-shape 重编译。

   建议：后续 PR 考虑将同样的 runtime arg 优化应用到 `l2norm_fwd`，或在 `l2norm_fwd` 中也切换到新 kernel。不阻塞本 PR。

2. **测试注释与实际 dispatch 逻辑不一致** [P3]

   `test_parity_dispatch_branches` 的注释写 `# T<=128 走 BT-tiled kernel；其余走 (T,) per-row kernel`，但实际 dispatch 条件是 `D <= 512`（BT-tiled）vs `D > 512`（per-row），与 T 无关。

   位置：`test_l2norm.py` 第 232 行附近

3. **BUILD 文件末尾缺少换行符** [P3]

   `test/BUILD` 最后一行 `)` 后没有尾部换行（diff 显示 `\ No newline at end of file`）。

### 整体评价

这是一个干净、聚焦的性能优化 PR。核心思路（rsqrt 替代 div、fuse q/k、T 作为 runtime arg）合理且有实测数据支撑（17x 加速）。代码改动范围小（4 文件，+201/-4），测试覆盖充分，commit 原子性好。原 `l2norm_fwd` 保留不变，不影响其他调用方。

✅ **LGTM ready to ci**

**Comment by LLLLKKKK** [comment]
🤖 AI Code Review — PR #920

## PR 概述

**Title:** `amd qwen35 optimize fused_l2norm_qk`
**Author:** hxy0118
**规模:** 4(GitHub) files, +201/-4

### 核心目标

将 `chunk.py` 中对 q、k 的两次独立 `l2norm_fwd` 调用合并为一次 `fused_l2norm_qk` kernel launch，同时用 `rsqrt + mul` 替代 `div`（AMD/NV 上 div 是多指令 rcp+NR，广播到 BT*BD 成为瓶颈），并将 T/D 改为 runtime arg 消除 per-shape 重编译。MI308X 上 9B prefill 从 842us 降至 49us（17x）。

---

## 改动逻辑拆解

### GitHub 开源仓库变更（主要代码）

#### 1. l2norm.py — 新增 fused kernel 及 dispatcher

新增两个 Triton kernel：

- `fused_l2norm_qk_kernel`：BT-tiled 路径（D<=512），T/D 为 runtime arg，BT/BD 为 constexpr。每个 program 处理 BT 行的 q 和 k，用 `rsqrt(sum(x*x) + eps) * x` 替代原来的 `x / sqrt(var + eps)`。
- `fused_l2norm_qk_kernel1`：per-row 路径（D>512），每个 program 处理一行 q 和一行 k。

新增 dispatcher `fused_l2norm_qk(q, k, eps, output_dtype)`：
- 校验 q/k shape 和 dtype 一致
- D<=512 走 BT-tiled kernel（BT=16, num_warps=2）
- D>512 走 per-row kernel（num_warps=8）
- 返回 `(q_out, k_out)` tuple

关键设计决策：T 和 D 不再是 `tl.constexpr`，避免不同 batch size 触发重编译（原 `l2norm_fwd_kernel` 的 T 是 constexpr，每个新 T 值都要编译一次，70ms 编译 vs <1ms 执行）。

#### 2. chunk.py — 调用方适配

import 从 `l2norm_fwd` 改为 `fused_l2norm_qk`，`forward()` 中两行 `q = l2norm_fwd(q); k = l2norm_fwd(k)` 合并为 `q, k = fused_l2norm_qk(q, k)`。

#### 3. test/test_l2norm.py — 新增 TestFusedL2NormQK 测试类

- `_check_parity`：对比 fused 结果与两次独立 `l2norm_fwd` 的结果
- `test_parity_qwen3_next_shapes`：D=128, T 从 16 到 65536
- `test_parity_dispatch_branches`：覆盖 D<=512 和 D>512 两条路径，bf16/fp16/fp32 三种 dtype
- `test_shape_dtype_assert`：校验 shape/dtype 不匹配时的 assert

#### 4. test/BUILD — 新增 ROCm 测试 target

新增 `test_l2norm_rocm`，复用 `test_l2norm.py`，指定 `tags = ["rocm"]` 和 `exec_properties = {'gpu':'MI308X'}`。

---

## Checklist 检查结果

### 通用原则

#### 软件工程原则
| 检查项 | 结果 |
|--------|------|
| SRP | ✅ 单一职责：fused kernel + dispatcher + 调用方适配 |
| OCP | ✅ 新增 fused 函数，未修改原 `l2norm_fwd` 逻辑 |
| LSP | ✅ 不涉及继承 |
| ISP | ✅ 接口精简 |
| DIP | ✅ 不涉及 |
| DRY | ✅ fused kernel 内 q/k 处理逻辑相同但操作不同数据，无需抽取 |
| KISS | ✅ 实现直接，无过度抽象 |
| YAGNI | ✅ 只做了需要的优化 |

#### 架构审视
| 检查项 | 结果 |
|--------|------|
| 抽象边界 | ✅ fused kernel 放在 l2norm.py 中，与原有 kernel 同层 |
| 依赖方向 | ✅ chunk.py → l2norm.py，方向不变 |
| 状态完整性 | ✅ 无状态变更 |
| 错误语义 | ✅ shape/dtype 不匹配时 assert 失败，D 过大时 RuntimeError |
| 可观测性 | ✅ 不涉及 |
| 可演进性 | ✅ 原 `l2norm_fwd` 保留，不影响其他调用方 |
| 可运维性 | ✅ 不涉及 |

#### 测试
| 检查项 | 结果 |
|--------|------|
| 新功能有对应测试 | ✅ TestFusedL2NormQK 覆盖多种 shape/dtype/dispatch 路径 |
| 删除的测试有等价替代 | ✅ 未删除测试 |
| 边界 case 覆盖 | ✅ T=16（小）到 T=65536（大），D=128/256/1024，三种 dtype |
| 分布式改动有多卡测试 | ✅ 不涉及分布式 |

#### 代码质量与文档
| 检查项 | 结果 |
|--------|------|
| 无关改动分离 | ✅ 所有改动围绕 fused l2norm |
| mega-PR 拆分 | ✅ 单一功能 |
| Commit 原子性 | ✅ 单 commit，自包含 |
| Commit message 准确性 | ✅ "optimize fused_l2norm_qk: rsqrt+mul, BT-tiled for all T" 准确描述 |
| PR description 说明动机和设计 | ✅ 包含优化原理和性能数据 |
| 日志频率控制 | ✅ 不涉及 |

### 领域检查

#### A. 兼容性与配置 — 全部 ✅
#### B. 正确性与逻辑 — 全部 ✅

kernel 内 `rsqrt(sum(x*x) + eps) * x` 与原 `x / sqrt(var + eps)` 数学等价。boundary mask `rows < T` 和 `cols < D` 正确处理边界。`other=0.0` 确保 padding 区域不影响 sum。

#### C. 线程安全与并发 — 全部 ✅
#### D. 性能 — 全部 ✅

核心优化点合理：rsqrt 是单条硬件指令，避免 div 的多指令开销；T 作为 runtime arg 消除重编译；fuse q/k 减少一次 kernel launch。

#### E. 分布式 — 全部 ✅（不涉及）
#### F. 跨平台（ROCm/ARM） — 全部 ✅

新增 ROCm 测试 target，kernel 使用标准 Triton 语法，跨平台兼容。

#### G. 语言与框架特有 — 全部 ✅
#### H. 测试与 CI — 全部 ✅
#### I. 代码质量 — 全部 ✅

---

## Review 意见

### 问题

1. **`l2norm_fwd` 中旧 BT-tiled kernel 仍有 T 作为 constexpr 的重编译问题** [P2]

   `fused_l2norm_qk_kernel` 已将 T/D 改为 runtime arg 解决了重编译问题，但原 `l2norm_fwd` 中的 `l2norm_fwd_kernel` 仍然将 T 声明为 `tl.constexpr`（第 60 行），且 `l2norm_fwd` 仍被 `L2NormFunction.forward` 调用。如果有其他路径（非 fused）调用 `l2norm_fwd`，仍会遇到 per-shape 重编译。

   建议：后续 PR 考虑将同样的 runtime arg 优化应用到 `l2norm_fwd`，或在 `l2norm_fwd` 中也切换到新 kernel。不阻塞本 PR。

2. **测试注释与实际 dispatch 逻辑不一致** [P3]

   `test_parity_dispatch_branches` 的注释写 `# T<=128 走 BT-tiled kernel；其余走 (T,) per-row kernel`，但实际 dispatch 条件是 `D <= 512`（BT-tiled）vs `D > 512`（per-row），与 T 无关。

   位置：`test_l2norm.py` 第 232 行附近

3. **BUILD 文件末尾缺少换行符** [P3]

   `test/BUILD` 最后一行 `exec_properties = {'gpu':'MI308X'},` 后的 `)` 没有尾部换行（diff 显示 `\ No newline at end of file`）。

### 小问题

无其他 nit。

### 整体评价

这是一个干净、聚焦的性能优化 PR。核心思路（rsqrt 替代 div、fuse q/k、T 作为 runtime arg）合理且有实测数据支撑（17x 加速）。代码改动范围小（4 文件，+201/-4），测试覆盖充分，commit 原子性好。原 `l2norm_fwd` 保留不变，不影响其他调用方。

✅ **LGTM ready to ci**

File: rtp_llm/models_py/triton_kernels/fla/l2norm.py [hunk 1]
```diff
@@ -124,6 +187,70 @@ def grid(meta):
...
+    assert (
+        q.shape == k.shape
+    ), f"fused_l2norm_qk expects q and k to share shape, got {q.shape} vs {k.shape}"
+    assert (
+        q.dtype == k.dtype
+    ), f"fused_l2norm_qk expects q and k to share dtype, got {q.dtype} vs {k.dtype}"
+    q_shape_og = q.shape
+    q_flat = q.view(-1, q.shape[-1])
+    k_flat = k.view(-1, k.shape[-1])
+    T, D = q_flat.shape
```
Comment (ZhangZhiPku) [inline rtp_llm/models_py/triton_kernels/fla/l2norm.py]: 需要修改 T -> tokens, D -> hidden_dim

File: rtp_llm/models_py/triton_kernels/fla/l2norm.py [hunk 2]
```diff
@@ -124,6 +187,70 @@ def grid(meta):
...
+            eps,
+            T,
+            D,
+            BT=BT,
+            BD=BD,
+            num_warps=2,
+            num_stages=1,
+        )
+    else:
+        fused_l2norm_qk_kernel1[(T,)](
```
Comment (ZhangZhiPku) [inline rtp_llm/models_py/triton_kernels/fla/l2norm.py]: fused_l2norm_qk_kernel1 需要命名为 fused_l2norm_qk_kernel

**Comment by LLLLKKKK** [comment]
## AI Code Review — PR #920

### PR 概述

**Title:** `amd qwen35 optimize fused_l2norm_qk`
**Author:** hxy0118
**规模:** 4 files, +206/-4

为 AMD GPU 优化 Qwen3.5 的 L2 norm 计算：将 q/k 两次独立 `l2norm_fwd` 合并为 fused kernel `fused_l2norm_qk`，同时用 `rsqrt + mul` 替代 `div`。仅在 AMD 平台启用。

---

### Review 意见

1. **`fused_l2norm_qk` 未被 `input_guard` 保护** [P2]
   `l2norm_fwd` 经 `L2NormFunction` 有 `@input_guard` 保护，但 `fused_l2norm_qk` 是裸函数，`q.view(-1, ...)` 对非连续 tensor 会报错。当前调用链有保护，但作为公开 API 建议加 `q = q.contiguous()` 或用 `reshape`。

2. **BT=8 硬编码** [P2]
   BT-tiled kernel 的 `BT=8` 是硬编码值，对大 T 场景可能不是最优。建议补充 benchmark 依据或考虑参数化。

3. **BUILD 文件缺少末尾换行** [P2]

4. **PR description 为空** [P3]
   建议补充性能对比数据和设计动机。

### 亮点

- rsqrt+mul 替代 div 是标准且有效的 GPU 优化
- `is_amd` 守卫确保 NVIDIA 路径不受影响
- 测试覆盖全面：多 shape、多 dtype、dispatch 分支、错误输入

### 结论

无 P0/P1 问题。P2 建议后续改进但不阻塞合入。

lgtm ready to ci

**Comment by LLLLKKKK** [comment]
## AI Code Review — PR #920

### PR 概述

**Title:** `amd qwen35 optimize fused_l2norm_qk`
**Author:** hxy0118
**规模:** 4 files, +207/-4

为 AMD GPU 优化 Qwen3.5 的 L2 norm 计算：将 q/k 两次独立 `l2norm_fwd` 合并为 fused kernel，用 rsqrt+mul 替代 div。MI308X 实测 842us -> 49us (17x)。

### Verdict: LGTM

无正确性问题。AMD-only gating 干净，测试覆盖充分。

### Issues

| Severity | Title | Location |
|----------|-------|----------|
| P2 | BT=8 hardcoded — consider autotune | `l2norm.py:230` |
| P2 | Kernel parameter ordering mismatch | `l2norm.py:96` |
| P3 | BUILD file missing trailing newline | `test/BUILD:61` |
| P3 | Test comment references old dispatch logic | `test/test_l2norm.py:237` |

**P2: BT=8 hardcoded** — BT=8 is a good default but `triton.autotune` with a few BT candidates (4, 8, 16) could find better tile sizes across different T ranges.

**P2: Parameter ordering** — `fused_l2norm_qk_kernel` signature has `(q, k, q_out, k_out, D, BD, eps)` but call site uses `eps=eps` keyword before `D=hidden_dim`. Works but confusing to read.

**P3: Trailing newline** — BUILD file missing newline at EOF.

**P3: Stale comment** — Test comment says "T<=128 gate" but actual dispatch uses `hidden_dim <= 512`.

### Strengths

- 17x speedup with clear benchmarks on MI308X
- rsqrt+mul optimization well-motivated with hardware rationale in code comments
- Comprehensive test coverage (parity, shapes, dtypes, assertions)
- Clean AMD-only gating via `is_amd` flag

### CI Status

Latest CI run [38277699](https://code.alibaba-inc.com/foundation_models/RTP-LLM/ci/jobs?pipelineId=1346&pipelineRunId=38277699&createType=yaml) FAILED (old SHA `516b691316`, not current head). Previous run 38277688 SUCCESS.

> Note: PR is currently **closed** (not merged).
